include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -724,10 +744,16 @@ class discard_iterator
 #    define _ONEDPL_CONSTEXPR_FIX constexpr
 #endif
 
+    _ONEDPL_CONSTEXPR_FIX difference_type
+    operator-(const discard_iterator& __it) const
+    {
+        return __my_position_ - __it.__my_position_;
+    }
+
     _ONEDPL_CONSTEXPR_FIX bool
     operator==(const discard_iterator& __it) const
     {
-        return __my_position_ - __it.__my_position_ == 0;
+        return *this - __it == 0;
```

Comment:
Overall comment: one of the main ideas of refactoring for iterators was to make them fulfill the C++ Iterator named requirement (including subcategories such as ForwardIterator, RandomAccessIterator and so forth). In the current implementation we support RandomAccess only but it seems for us to support of them relatively cheap.

We recently did such a refactoring for tranform_iterator API. Please look there for the inspiration. I believe that's the one you'd need: #172

Reply:
@rarutyun, thank you for the pointer.  The `operator==` seems to be the only one that required an update to support the other iterator concepts.  The `discard_iterator` class isn't templated, so I don't think `iterator_category` can be customized as it is for `transform_iterator`, but I don't think this is an issue.  Please let me know if I'm missing something.